keps/sig-auth/5284-constrained-impersonation/README.md

```diff
@@ -0,0 +1,1286 @@
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentications.k8s.io
+    resource: uids
+    name: someUID
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer
+
+#### Header with prefix `Impersonate-Extra-` is set
+
+A subjectaccessreview
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentications.k8s.io
+    resource: userextras
+    subresource: extraKey
+    name: extraValue
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer
+
+### Verb `impersonate:serviceaccount`
+
+Same as legacy impersonation, when the request has the header of `Impersonate-User`, and the value of the header
+has a prefix of `system:serviceaccount:`, apiserver will check verb `impersonate:serviceaccount` with the authorizer
+using the subjectaccessreview:
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentications.k8s.io
+    resource: serviceaccounts
+    name: serviceaccount-name
+    namespace: serviceaccount-namespace
+    verb: impersonate:serviceaccount
+  user: impersonator
+```
+
+### Verb `impersonate:scheduled-node`
+
+This is a special verb to check when two conditions are met:
+1. The impersonator is impersonating a node by setting the header `Impersonate-User` and the value has a prefix
+of `system:node:` on the request.
+2. The user info of the impersonator has an extra with key `authentication.kubernetes.io/node-name`, and the value
+should be the same as the value in the request header of `Impersonate-User` after removing prefix `system:node:`.
+It indicates that the impersonator is running on the same node it is impersonating.
+
+If the two conditions are both met, the following subjectaccessreview will be sent to the authorizer:
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentications.k8s.io
+    resource: nodes
+    verb: impersonate:scheduled-node
+  user: impersonator
+```
+If one of the conditions is not met, verb `impersonate:scheduled-node` will not be checked and verb
+`impersonate:user-info` will be checked instead. For instance, if condition 2 is not met, the following
+subjectaccessreview will be sent.
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentications.k8s.io
+    resource: users
+    name: system:node:node1
+    verb: impersonate:user-info
+  user: impersonator
+```
+
+### Auditing
+
+An audit event is recorded already for any allowed impersonation request. To record the reason why the impersonation
+is allowed, an annotation `allowed-impersonation-verbs` will be added. The value will be the list of impersonation
+related verbs checked. For example, `allowed-impersonation-verbs: impersonate:scheduled-node,impersonate-on:list`
```

Comment:
I don't think we need the `impersonate-on:list` bit because that can be inferred from the request itself right?